From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: quh4gko8 <88831734+quh4gko8@users.noreply.github.com>
Date: Tue, 20 Feb 2024 04:05:46 +0000
Subject: [PATCH] Start child processes in JITless mode based on OS
 configurations

This is the current approach prior to Android 17 for starting isolated
processes in JITless mode, specifically when dynamic code loading in
memory is restricted, that is, when execmem is blocked in SELinux. This
implementation is kept behind a config-specfic feature flags that's
parsed and appended into per-process command line flags, where jitless
mode flags is set upon each isolated process initialized with ART.
---
 .../library_loader/LibraryLoaderHooks.java    | 54 +++++++++++++++++++
 base/base_library_loader_java_ext_deps.gni    |  1 +
 2 files changed, 55 insertions(+)

diff --git a/base/android/java/src/org/chromium/base/library_loader/LibraryLoaderHooks.java b/base/android/java/src/org/chromium/base/library_loader/LibraryLoaderHooks.java
index b1a621f62770a..4630221709891 100644
--- a/base/android/java/src/org/chromium/base/library_loader/LibraryLoaderHooks.java
+++ b/base/android/java/src/org/chromium/base/library_loader/LibraryLoaderHooks.java
@@ -2,16 +2,70 @@ package org.chromium.base.library_loader;
 
 import android.content.Context;
 
+import org.chromium.base.CommandLine;
+import org.chromium.base.Log;
+import org.chromium.base.command_line.VanadiumCommandLineUtils;
 import org.chromium.base.config.VanadiumConfigBridge;
 
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+
+import app.vanadium.config.VanadiumConfParser;
+import app.vanadium.ext.CustomOSApis;
+
+import static app.vanadium.config.proto.VanadiumConfigProto.Flag.FlagType.DISABLED_FEATURE;
+
 class LibraryLoaderHooks {
 
+    private static final String TAG = "LibraryLoaderHooks";
+
     static void onSetLibraryProcessType(Context appContext, @LibraryProcessType int type) {
         VanadiumConfigBridge.initConfigs(appContext, type);
     }
 
     static void onBeforeCommandLineSwitchLocked(Context appContext, @LibraryProcessType int type) {
         VanadiumConfigBridge.applyFlagFromConfigs(appContext, type);
+        switch (type) {
+            case LibraryProcessType.PROCESS_BROWSER:
+            case LibraryProcessType.PROCESS_WEBVIEW:
+            case LibraryProcessType.PROCESS_WEBVIEW_NONEMBEDDED:
+                return;
+            case LibraryProcessType.PROCESS_CHILD:
+            case LibraryProcessType.PROCESS_WEBVIEW_CHILD:
+                if (CommandLine.getInstance() != null) {
+                    if (!VanadiumConfParser.getFlagsFor(
+                            DISABLED_FEATURE).containsKey("CheckOsConfigsForIsolatedProcessV1")) {
+                    } else if (type == LibraryProcessType.PROCESS_CHILD
+                            && CustomOSApis.isMemoryDclRestricted() != null) {
+                        Log.i(TAG, "CheckOsConfigsForIsolatedProcessV1: disabled " +
+                                "by presence of valid isMemoryDclRestricted method");
+                        break;
+                    } else if (CustomOSApis.isWebViewDclRestricted() != null) {
+                        Log.i(TAG, "CheckOsConfigsForIsolatedProcessV1: disabled " +
+                                "by presence of valid isWebViewDclRestricted method");
+                        break;
+                    } else {
+                        Log.i(TAG, "CheckOsConfigsForIsolatedProcessV1: " +
+                                "disabled by config app");
+                        break;
+                    }
+                    final Boolean isExecmemBlocked = CustomOSApis.isExecmemBlocked();
+                    if (isExecmemBlocked == null || !isExecmemBlocked) {
+                        break;
+                    }
+
+                    String jsFlagsName = "js-flags";
+                    LinkedHashSet<String> jsFlags = new LinkedHashSet<>(
+                            VanadiumCommandLineUtils.getCommaDelimitedSwitchValue(jsFlagsName));
+                    jsFlags.add("--jitless");
+                    jsFlags.add("--wasm_jitless");
+                    VanadiumCommandLineUtils.setCommaDelimitedSwitchValue(
+                            jsFlagsName, new ArrayList<>(jsFlags));
+                }
+                break;
+            default:
+                throw new IllegalStateException();
+        }
     }
 
     static void onSubsequentCommandLineSwitchLockedCheck(Context appContext, @LibraryProcessType int type) {
diff --git a/base/base_library_loader_java_ext_deps.gni b/base/base_library_loader_java_ext_deps.gni
index 8300469435f39..db540a25dde69 100644
--- a/base/base_library_loader_java_ext_deps.gni
+++ b/base/base_library_loader_java_ext_deps.gni
@@ -4,4 +4,5 @@
 
 base_library_loader_java_ext_deps = [
   "//vanadium/android_config/proto:browser_config_parser_java",
+  "//vanadium/ext/custom_os:custom_os_apis_java",
 ]
